src/librustc/middle/const_eval.rs

```diff
@@ -607,6 +697,71 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
     Ok(result)
 }
 
+fn resolve_trait_associated_const<'a, 'tcx: 'a>(tcx: &'a ty::ctxt<'tcx>,
+                                                ti: &'tcx ast::TraitItem,
+                                                trait_id: ast::DefId,
+                                                ref_id: ast::NodeId)
+                                                -> Option<&'tcx Expr>
+{
+    let rcvr_substs = ty::node_id_item_substs(tcx, ref_id).substs;
```

Comment:
Good point; I'm not really sure. The `traits` code is probably the main bit that I don't understand as well as I should here, so here (and even more so in `rustc_typeck::check::compare_method`) there may be a couple of details that are just superstitiously copied.

One thing that concerns me is that maybe we could have a path like `<<T>::U as Trait>::CONST` or `<[T; <U>::N] as Trait>::CONST` (or the nested UFCS could be in the trait, too). The former is disallowed right now because `<T>::U` can't be used to refer to an associated type; I don't know if that ever will be allowed in the future. The latter is of course one of the things I don't have working yet, but it seems like it could be done.

As for moving `const_eval` to `rustc_typeck`, probably the main obstacle there is that `rustc_trans` uses it. It's kind of awkward because `middle::const_eval` and `trans::consts` kind of duplicate some of each others' logic, but the former is more broadly used. But I think that `const_eval` is doing some things wrong that are done right in `trans::consts`. E.g. in `const_eval` all floats are treated as `f64`, and `compare_const_vals` treats them as totally ordered.